velox/exec/Window.cpp

```diff
@@ -137,12 +137,47 @@ void Window::createWindowFunctions(
         operatorCtx_->pool(),
         &stringAllocator_));
 
-    windowFrames_.push_back(
-        {windowNodeFunction.frame.type,
-         windowNodeFunction.frame.startType,
-         windowNodeFunction.frame.endType,
-         fieldArgToChannel(windowNodeFunction.frame.startValue),
-         fieldArgToChannel(windowNodeFunction.frame.endValue)});
+    auto createFrameChannelArg =
+        [&](const core::TypedExprPtr& frame) -> FrameChannelArg {
+      if (auto constant = constantArg(frame)) {
+        auto constantOffset =
+            constant->template as<ConstantVector<int64_t>>()->valueAt(0);
+        VELOX_USER_CHECK_GE(
+            constantOffset, 1, "k in frame bounds must be at least 1");
+        return {kConstantChannel, constant};
+      } else {
+        return {
+            fieldArgToChannel(frame).value(),
+            BaseVector::create(BIGINT(), 0, operatorCtx_->pool())};
+      }
+    };
+
+    bool hasKBasedStartBound =
+        (windowNodeFunction.frame.startType ==
+             core::WindowNode::BoundType::kPreceding ||
+         windowNodeFunction.frame.startType ==
+             core::WindowNode::BoundType::kFollowing);
+    bool hasKBasedEndBound =
+        (windowNodeFunction.frame.endType ==
+             core::WindowNode::BoundType::kPreceding ||
+         windowNodeFunction.frame.endType ==
+             core::WindowNode::BoundType::kFollowing);
+    if (windowNodeFunction.frame.type == core::WindowNode::WindowType::kRange &&
+        (hasKBasedStartBound || hasKBasedEndBound)) {
+      VELOX_NYI("k preceding/following is only supported in ROWS mode");
+    }
+
+    windowFrames_.push_back({
+        windowNodeFunction.frame.type,
+        windowNodeFunction.frame.startType,
+        windowNodeFunction.frame.endType,
+        (hasKBasedStartBound ? std::make_optional(createFrameChannelArg(
```

Comment:
Assuming that frame.startValue is null unless bound type is kPreceding or kFollowing (we should check that in WindowNode's constructor), there is not need for hasKBasedEndBound and similar variables. Just call createFrameChannelArg and let it return std::nullopt if frame.startValue is null.